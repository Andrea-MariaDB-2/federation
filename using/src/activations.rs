//! Methods to select which of a set of implementations to activate based on a schema's
//! `Request`s.

use crate::{Find, Request, Schema, Version};

/// An Activation references a request and contains the min and max versioned implementations
/// available in the implementation registry which can satisfy that request, or None if no
/// matching implementations were available.
#[derive(Debug, PartialEq, Eq)]
pub struct Activation<'schema, 'impls, T> {
    pub request: &'schema Request,
    pub min: Option<(&'impls Version, &'impls T)>,
    pub max: Option<(&'impls Version, &'impls T)>,
}

impl<'a> Schema<'a> {
    /// Given a set of implementations, take all requests in the document and
    /// return an iterator over `Activation`s containing the min- and max-
    /// versioned implementations which will satisfy the request if any such
    /// implementations are available.
    pub fn activations<'impls, T, F>(&'impls self, implementations: &'impls F) -> impl Iterator<Item=Activation<'a, 'impls, T>> + 'impls
        where
            'impls: 'a,
            T: 'impls,
            F: Find<T>
    {
        self.using.iter()
            .map(move |req| (
                req,
                implementations.find_min_req(req),
                implementations.find_max_req(req),
            ))
            .map(|(request, min, max)|
                Activation { request, min, max }
            )
    }
}

#[cfg(test)]
mod tests {
    use graphql_parser::ParseError;
    use crate::*;
    
    #[test]
    fn it_iterates_over_activations() -> Result<(), ParseError> {
        let implementations = Implementations::new()
            .provide("https://spec.example.com/A", Version(1, 2), "impl A v1.2".to_owned())
            .provide("https://spec.example.com/B", Version(1, 2), "impl B v1.2".to_owned());

        let mut expected = vec![
r#"Activation {
    request: Request {
        spec: Spec {
            identity: "https://spec.example.com/A",
            default_prefix: "A",
            version: Version(
                1,
                0,
            ),
        },
        prefix: "A",
        position: Pos(3:17),
    },
    min: Some(
        (
            Version(
                1,
                2,
            ),
            "impl A v1.2",
        ),
    ),
    max: Some(
        (
            Version(
                1,
                2,
            ),
            "impl A v1.2",
        ),
    ),
}"#,
r#"Activation {
    request: Request {
        spec: Spec {
            identity: "https://spec.example.com/unknown",
            default_prefix: "unknown",
            version: Version(
                1,
                0,
            ),
        },
        prefix: "unknown",
        position: Pos(4:17),
    },
    min: None,
    max: None,
}"#
        ];
        expected.reverse();

        Schema::parse(r#"
            schema
                @using(spec: "https://spec.example.com/A/v1.0")
                @using(spec: "https://spec.example.com/unknown/v1.0")
            {
                query: Query
            }
        "#)?
            .activations(&implementations)
            .for_each(|req|
                assert_eq!(expected.pop().unwrap(), format!("{:#?}", req)));
        
        Ok(())
    }
}