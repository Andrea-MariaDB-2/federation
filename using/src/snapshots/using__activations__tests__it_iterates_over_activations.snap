---
source: using/src/activations.rs
expression: "{\n    let implementations =\n        Implementations::new().provide(\"https://spec.example.com/A\",\n                                       Version(1, 2),\n                                       \"impl A v1.2\".to_owned()).provide(\"https://spec.example.com/B\",\n                                                                         Version(1,\n                                                                                 2),\n                                                                         \"impl B v1.2\".to_owned());\n    let schema =\n        Schema::parse(r#\"\n                schema\n                    @core(using: \"https://lib.apollo.dev/core/v0.1\")\n                    @core(using: \"https://spec.example.com/A/v1.0\")\n                    @core(using: \"https://spec.example.com/unknown/v1.0\")\n                {\n                    query: Query\n                }\n                \"#)?;\n    let activations =\n        schema.activations(&implementations).map(|(req, find)|\n                                                     (req,\n                                                      find.collect::<Vec<_>>())).collect::<Vec<_>>();\n    format!(\"{:#?}\", activations)\n}"
---
[
    (
        Request {
            spec: Spec {
                identity: "https://lib.apollo.dev/core",
                name: "core",
                version: Version(
                    0,
                    1,
                ),
            },
            name: "core",
            position: Pos(3:21),
        },
        [],
    ),
    (
        Request {
            spec: Spec {
                identity: "https://spec.example.com/A",
                name: "A",
                version: Version(
                    1,
                    0,
                ),
            },
            name: "A",
            position: Pos(4:21),
        },
        [
            (
                Version(
                    1,
                    2,
                ),
                "impl A v1.2",
            ),
        ],
    ),
    (
        Request {
            spec: Spec {
                identity: "https://spec.example.com/unknown",
                name: "unknown",
                version: Version(
                    1,
                    0,
                ),
            },
            name: "unknown",
            position: Pos(5:21),
        },
        [],
    ),
]
